### PR TITLE
some useful plotting and numpy matrix functions and a notebook

### PR DESCRIPTION
**Pull request by lschwetlick**
Done:
- A notebook explaining how to use and visualize Pelita stuff in a notebook
- some helper functions in utils

Todo:
- We still need to come up with a useful but simple usecase for numpy, that isnt just visualization
- We need to fix the notebook's layout loading (with the `load_builtin_layout` function)
- testing the stuff that isnt strictly visualization?

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -19,3 +19,98 @@ def walls_to_graph(walls):
                         # this is a genuine neighbor, add an edge in the graph
                         graph.add_edge((x, y), neighbor)
     return graph
+
+
+import numpy as np
+
+def draw_coords(coords, ax, params={}):
+    """
+    Takes a list of coordinates and
```
Comment (otizonaizit) [inline utils.py]: the sentence is not finished?

File: utils.py [hunk 2]
```diff
@@ -19,3 +19,98 @@ def walls_to_graph(walls):
...
+    mat = np.zeros((maze_height, maze_width))
+    # get seperate lists of x and y coordinates from a list of tuples
+    wall_x, wall_y = zip(*bot.walls)
+    # matrices have y,x indexing!
+    mat[wall_y, wall_x] = 1
+    return mat
+
+def draw_game_elements(bot, ax):
+    """
+    draws game state as scatter elements given a bot object.
```
Comment (otizonaizit) [inline utils.py]: Capital `d` in `draws`

File: utils.py [hunk 3]
```diff
@@ -19,3 +19,98 @@ def walls_to_graph(walls):
                         # this is a genuine neighbor, add an edge in the graph
                         graph.add_edge((x, y), neighbor)
     return graph
+
+
+import numpy as np
```
Comment (otizonaizit) [inline utils.py]: the import should go at the top of the file?

**Comment by otizonaizit** [comment]
By the way, the notebook is quite nice indeed. Yes, the thing with the layout needs to be fixed so that the students don't ever get to see the "internal" representation. I thought @Debilski addressed part of the problem already in https://github.com/ASPP/pelita/pull/657

**Comment by lschwetlick** [comment]
> I thought @Debilski addressed part of the problem already in ASPP/pelita#657

Yup, just that was after I made the notebook. I added it now.

> So, maybe the whole thing should stay in the notebook?

Yeah, I sort of agree. Then also the code is right there when you scroll through the notebook, which is nice! Also, yes, like you said you can convert it however you like depending what you want to do... so its probably better off as an example.

I was thinking of making another notebook that could explain the enemy noise. That would maybe also solve [issue 13](https://github.com/ASPP/pelita_template/issues/13). What do you think?

**Comment by lschwetlick** [comment]
Another thing was what you guys were talking about in the car: it would be nice to have a function that returns a new gamestate given a move. Then you could plot, simulate a move and visualize again. I assume that's what @Debilski  is working on [here](https://github.com/ASPP/pelita/pull/650)?

**Comment by otizonaizit** [comment]
@lschwetlick  : Good work! @Debilski : could you have a look also and merge if you are convinced? I never use notebooks, so I can't really judge the technical side of it. The code is good in my opinion. The plots are quite small in my browser and I almost can not distinguish the bots, but it may be just my settings?

**Comment by lschwetlick** [comment]
I made everything a bit larger now :)

**Comment by otizonaizit** [comment]
thanks, plots are much more readable now!
@Debilski : please look at this and merge or comment. Do you think it's clear enough to explain what the noiser does?

**Comment by Debilski** [comment]
Uhm, that is not how the noiser works :) 
image


This is more accurate:
image

**Comment by Debilski** [comment]
> Another thing was what you guys were talking about in the car: it would be nice to have a function that returns a new gamestate given a move. Then you could plot, simulate a move and visualize again. I assume that's what @Debilski is working on [here](https://github.com/ASPP/pelita/pull/650)?

Yes, but I am not sure yet, if this is the right approach. I’ll still have to think about it more.

A few more notes:

- Don’t do `import pelita as pl`, this makes things even more confusing
- I am not sure, we want to add the noise demo to the repo. I think it should be something for the students to write in their own preferred style, if they want it. (They can easily figure it out from the scatter and imshow examples.)
- I would prefer that we keep the order of (x, y) in a numpy array. It should be reversed in the plotting routines.
- width and height have a helper function in layouts.py but I am not sure if we have a consensus that students should use them
- `draw_game_elements` → `draw_bot`?

**Comment by lschwetlick** [comment]
> Uhm, that is not how the noiser works :)

Really, because then maybe we do need to talk again about the wording in the readme:

> bot.enemy[0].is_noisy if the enemy bot is located more than 5 squares away from your bot, then its position bot.enemy[0].position will not be exact. A uniformly distributed noise between -5 and +5 squares will be added to it instead. The distance of your bot from the enemy bot is measured in grid space, i.e. if your bot is in (Bx, By) and the enemy bot is in (Ex, Ey), the distance will be abs(Bx-Ex) + abs(By-Ey). An example of using the is_noisy property is given in demo05_basic_defender.py.

It makes it clear that the distance between the enemy and me measured in steps, but the despriction of the noise sounds like (x,y) just get +/-0 to 1 thrown on top of it.

> I am not sure, we want to add the noise demo to the repo

Hm, okay. I dont think it gives too much away and it might be nice to see...

> draw_game_elements → draw_bot

it draws more than just the bot, theres also food, walls etc, no?

> width and height have a helper function in layouts.py but I am not sure if we have a consensus that students should use them

Didnt realise. Should I include it?

**Comment by Debilski** [comment]
> it draws more than just the bot, theres also food, walls etc, no?

True, but since it takes a bot object as its argument, I find it a little more fitting. `print(bot)` gives you basically the same output. But maybe just name it `draw_game` then.

>> width and height have a helper function in layouts.py but I am not sure if we have a consensus that students should use them

> Didnt realise. Should I include it?

Nah, maybe not. (Students should only care about `pelita.utils`.) But you could just put all your helper function on top of the notebook and also define this one there.

In the afternoon I’ll spend some more time on how to document the noiser better.

**Comment by lschwetlick** [comment]
So just to be clear, I'll fix the things @Debilski suggested, but I wont fix the noise notebook unless you want to keep it. Should we scrap it or not?

**Comment by Debilski** [comment]
If you find the time, you can still clean it up a little, I think the consensus was that we’ll probably keep it in one way or the other. I’ll only have time on Wednesday though to look at it in detail again. Still, I’d focus more on the first one, to have that one look good and understandable. The noise notebook will then be easy to adapt.

**Comment by otizonaizit** [comment]
@lschwetlick : while you finish the cleanup, check out: https://github.com/ASPP/pelita/commit/d819a600b6ddd9a21c5d79b3c783795fab4a9af4 . With this you can pass directly the name of a built-in layout to `setup_test_game` . You can remove all of the layout parsing stuff: I will use this also in some other of the test_demoxx files, so that people see that this can be done.
It is much better now, because we don't need to expose the internals of pelita, nor the users ever get to see the internal representation of the layout

**Comment by lschwetlick** [comment]
Okay, I checked both of them and fixed the noise nb.

> I would prefer that we keep the order of (x, y) in a numpy array. It should be reversed in the plotting routines.

I thought about this a bit and I'm not sure I agree. I definitely see your point that it would be less ugly to keep (x,y) notation as long as possible. On the other hand it is valuable to be very aware of the difference between matrix indexing and cartesian coordinates. For me at least, this is regularly a cause for errors. Also, the way I think about it, I would expect my numpy matrix representation to be the right way round when I call print(array).

**Comment by Debilski** [comment]
> I'll change it if you guys really want, but I left it for now.

I’m the wrong person to decide here. I’m not able to memorise which matrix index is which anyway. (Left–right confusion but with rows and columns.)

**Comment by otizonaizit** [comment]
I think the (y, x) coordinate problem is spread all-over numpy and the scipy ecosystem already. Just look for example at the help for the `numpy.meshgrid` function.
I agree with @lschwetlick that I value more `print(array)` and `plt.imshow(array)` to show the thing in the right order out of the box, than having `array[x,y]` corresponding to `(x, y)` in the maze.
So, given that @Debilski declared is inability to decide, I think @lschwetlick wins.

@lschwetlick : can you post here a comment with text `merge?` when you think this PR can go in?

**Comment by lschwetlick** [comment]
merge?

**Comment by otizonaizit** [comment]
Can you fix the conflicts first?

**Pull request merged**